### DISCUSSION TITLE
A list of new functions/ideas

### DISCUSSION THREAD

**Discussion by mrozpara** [gillett-hernandez/Foreman2#8] [Ideas]
I'm thinking about the following list of enhancements/new functions:

1. Tabs - ability to work on several graphs
2. Quick access to saved graphs - as list
3. 'Done' flag for assemblers - to mark which buildings are already built
4. Labels - leaving notes on graph
5. List of all nodes with auto zoom function (on double click)
6. List of researched/not-researched technologies - enabling/disabling recipes/assemblers (currently it can be done only based on research level)
7. Display required technology for disabled recipes/buildings.
8. More user friendly selection of entities that produce energy (for example PY energy from Auog is quite tricky to find... at least for me)

1-4 - I've got it developed but code refactoring and bugfixing is required...
![image](https://user-images.githubusercontent.com/94693225/202572770-cd7f8037-57b8-4656-801f-b9b9084f5931.png)
![image](https://user-images.githubusercontent.com/94693225/202572860-5c088015-e503-4292-8b63-7a59de3486eb.png)

5-6 - I've got initial functions, some more is required

What do you think about it? Shell I refactor code and prepare pull requests?

**Comment by gillett-hernandez** [comment]
Hi, sorry for the extremely late reply.

1. i definitely want this implemented one way or another, however i had another idea that might inform how it's implemented. i want to have a subgraph / nodegroup feature similar to blender, where you can bundle nodes into a nodegroup that is represented as a single node with aggregated inputs and outputs. i also want this to be something that can be baked into the current graph file (.fjson file) or also split out into multiple files such that you could, for instance, have a collection of graphs representing various blocks in your factory, and then a master factory plan that references all of those graphs and connects them all up, while also transparently working with the calculation system so that the numbers are all correct. Likely this will end up being implemented as nodes having a layer id and only displaying nodes for the current layer, but having them all in existence even if they're not visible, with passthrough nodes being the connection method in and out of subgraphs.
2. would also be good, preferably implemented as a menu under File, i.e. Recently Opened Graphs
3. would be great and shouldn't be too hard, not sure about the visuals
4. would be good.
5. would be good, however we'd also want multiple search methods (recipe, item) and for it to display the number of inputs and outputs and the recipe or node type and building 
6. yes to this as well, i see you already have some progress made
7. seems like it might be tricky to determine which technologies would be most relevant to show to the user as for items specifically, multiple techs "unlock" an item through the recipes and buildings they unlock. but still this shouldn't be too hard, for instance Recipe Book already has this feature.
8. would definitely be useful, likely easiest as just as a toggle in the "add recipe" panel/window. that way it wouldn't have to be its own modal/window thing. though this would need to be done in tandem or after an improvement to displaying items in the add item browser. for instance adding hot molten salt in py shows a bunch of the same items and the only thing making them distinct is the required tech. we'd want to also show the buildings that can make this item or recipe, or the required technology, per 7

**Comment by gillett-hernandez** [comment]
also re:1 it might need special NodeGroup Supply and NodeGroup Sink nodes rather than just passthroughs, and those would get converted to Passthroughs when instantiated as a nodegroup. Also it might be hard to handle reconciling the constraints that the user might incidentally or intentionally have on inputs and outputs i.e. max temp, min temp for fluids, and also limited throughput supplies and sinks.

**Comment by mrozpara** [comment]
re:1 - regarding "block recipes" (as I'm calling it...) - I'm trying to design something similar to one you described.
The part of functionality is already implemented in original version of Foreman - see:
`public static JObject PrepPreset(Preset preset)` and Custom Preset
My approach: save graph as recipe in Custom json and later use it in "big graph" (some additional checks such as all source/output must be properly assigned).
I'm just implementing any idea that I need to have my Py graphs somehow managable - but ...

**Reply by gillett-hernandez** [reply]
That could work but it wouldn't be able to robustly handle all the kinds of graphs users can make that they might want to create as a subgraph. For instance any nonlinear response through the use of passthrough limited nodes wouldn't be modeled if we just encapsulated and abstracted a graph. Maybe it could be a toggle though, like `[ ] simple graph` and said toggle changes whether it's stored and calculated as a custom recipe or as the full graph.

**Comment by mrozpara** [comment]
there should be a verification:
- there are no unconnected outputs - everything should be linked to "output" or voided
- there are no under/over production
- everything has a proper source

I'm not sure how "limited passthrough" could affect subgraph? if you treat subgraph as an assembler (or block) it will just recalculate the number of "assemblers" and everything inside will be changed proportionally. If you would like to recalculate the number of assemblers inside subgraph ... hmm...

**Reply by gillett-hernandez** [reply]
A limited passthrough node would affect things the same way a belt bottleneck does in-game. Everything downstream of the bottleneck is limited by that input speed and everything upstream would be limited by the output speed

**Reply by gillett-hernandez** [reply]
To be honest I don't use limited passthrough nodes so it's not too much of an issue. we could just add that as a validation step, that no there are no limited passthrough nodes in the subgraph

**Comment by mrozpara** [comment]
IMHO: under/over production check will handle it.
Any way - I will  try to finish my code and share the results.

Regarding #3 "Done flag" - I've got it ready, however there are some "issues with the logic":
- shell "done flag" be reset in case of recalc/changed # of assemblers?
- shell graph be "locked" (calculations paused) in order to prevent change of number of assemblers?

in my version I don't care about it (I know how it's working...) but maybe it should be fixed? What do you suggest? (maybe we should split this discussion to several threads...)

Regarding PY and TURD (and changes in recipes) - any ideas how to handle it?

**Reply by gillett-hernandez** [reply]
I don't think the done flags should be reset when things change nor do i think the graph should be locked.

regarding the TURDs, I'm not sure yet. I'd need to research more how the beacons work in Foreman2 as that is how TURD effects are implemented in-game.

**Comment by gillett-hernandez** [comment]
On a separate note, I'm doing preliminary work on adding a Undo / Redo feature to the application, with configurable length in terms of how many undo and redo steps will be kept in memory. I can push it to a separate branch once I have something working so you can test it if you'd like. Or I can directly upload a build if you prefer.
